In [2]:
import pandas as pd
from dotenv import load_dotenv
from openai import OpenAI
from anthropic import Anthropic
from src.consts import *

load_dotenv(override=True)
openai = OpenAI()
anthropic = Anthropic() 

# Load the data files
oracle_df = pd.read_csv('ThePauperCube_oracle_with_pt.csv')
print(f"Loaded {len(oracle_df)} cards from oracle_df")
print(f"Columns available: {list(oracle_df.columns)}")
oracle_df.head()

Loaded 450 cards from oracle_df
Columns available: ['name', 'CMC', 'Type', 'Color', 'Color Category', 'Oracle Text', 'tags', 'MTGO ID', 'Power', 'Toughness']


name  CMC                      Type Color Color Category  \
0         Boros Elite    1  Creature - Human Soldier     W          White   
1     Deftblade Elite    1  Creature - Human Soldier     W          White   
2     Doomed Traveler    1  Creature - Human Soldier     W          White   
3      Elite Vanguard    1  Creature - Human Soldier     W          White   
4  Faerie Guidemother    1        Creature - Faerie      W          White   

                                         Oracle Text  \
0  Battalion — Whenever this creature and at leas...   
1  Provoke (Whenever this creature attacks, you m...   
2  When this creature dies, create a 1/1 white Sp...   
3                                                NaN   
4  Flying // Target creature gets +2/+1 and gains...   

                                             tags   MTGO ID  Power  Toughness  
0                                             NaN  120547.0    1.0        1.0  
1                                             NaN   18617.0    1.0        1.0  
2  GTC Update;token generator;WR tokens;WG tokens   42650.0    1.0        1.0  
3                                      EMA Update   60565.0    2.0        1.0  
4                                      ELD Update   78110.0    1.0        1.0

# Theme Validation
Let's check if we have enough cards available for each theme in our jumpstart cube.

# Deck Construction
Now let's test the deck construction function to build actual jumpstart decks from our themes.

In [13]:
# Test the deck construction function
from src.construct import construct_jumpstart_decks, analyze_deck_composition, print_deck_summary

# Build all jumpstart decks
print("🚀 Starting deck construction...")
deck_dataframes = construct_jumpstart_decks(oracle_df, target_deck_size=13)

# Print comprehensive summary
print_deck_summary(deck_dataframes)

🚀 Starting deck construction...
🏗️ CONSTRUCTING JUMPSTART DECKS

📦 Phase 1: Assigning dual-color cards to dual-color themes

🎯 Building Control deck (W/U)
  ✅ Added: Judge's Familiar (Score: 3.0)
  ✅ Added: Momentary Blink (Score: 2.3)
  ✅ Added: Skybridge Towers (Score: 1.0)
  📊 Phase 1 complete: 3/13 cards

🎯 Building Mill deck (U/B)
  ✅ Added: Soul Manipulation (Score: 1.0)
  ✅ Added: Waterfront District (Score: 1.0)
  ✅ Added: Dimir Guildmage (Score: 0.5)
  📊 Phase 1 complete: 3/13 cards

🎯 Building Aggro deck (B/R)
  ✅ Added: Fireblade Artist (Score: 3.3)
  ✅ Added: Body Dropper (Score: 1.3)
  ✅ Added: Blightning (Score: 1.3)
  ✅ Added: Tramway Station (Score: 1.0)
  📊 Phase 1 complete: 4/13 cards

🎯 Building Big Creatures deck (R/G)
  📊 Phase 1 complete: 0/13 cards

🎯 Building Tokens deck (G/W)
  📊 Phase 1 complete: 0/13 cards

🎯 Building Lifedrain deck (W/B)
  ✅ Added: Kingpin's Pet (Score: 2.0)
  ✅ Added: Gift of Orzhova (Score: 1.5)
  ✅ Added: Pillory of the Sleepless (Score: 

In [ ]:
# Import validation functions and run card uniqueness validation
from src.validation import validate_card_uniqueness, validate_deck_constraints, validate_jumpstart_cube, display_validation_summary

# Run the validation
validation_result = validate_card_uniqueness(deck_dataframes)

🔍 VALIDATING CARD UNIQUENESS
📊 VALIDATION RESULTS:
Total cards across all decks: 351
Unique cards used: 351
Duplicate cards found: 0

✅ VALIDATION PASSED!
All 351 cards are used exactly once.


In [ ]:
# Additional analysis using the imported validation functions
from src.validation import analyze_card_distribution

# Run the distribution analysis
distribution_analysis = analyze_card_distribution(deck_dataframes, oracle_df)


📈 CARD DISTRIBUTION ANALYSIS
📊 OVERALL STATISTICS:
Total cards available: 450
Total cards used: 351
Cards unused: 99
Usage rate: 78.0%

🎨 USAGE BY COLOR:
  White    :  49/ 67 cards ( 73.1%)
  Blue     :  57/ 66 cards ( 86.4%)
  Black    :  56/ 66 cards ( 84.8%)
  Red      :  57/ 67 cards ( 85.1%)
  Green    :  59/ 64 cards ( 92.2%)
  Colorless:  49/ 59 cards ( 83.1%)

🎯 DECK COMPLETENESS:
Complete decks (13 cards): 21
Incomplete decks: 9

Incomplete deck details:
  Mill: 12/13 cards
  Big Creatures: 1/13 cards
  Tokens: 6/13 cards
  Lifedrain: 12/13 cards
  Graveyard Value: 8/13 cards
  Equipment Aggro: 8/13 cards
  Angels: 11/13 cards
  Dragons: 10/13 cards
  Beasts: 10/13 cards

📋 UNUSED CARDS ANALYSIS:
Unused creatures: 46
Unused lands: 30
Unused spells: 23

Sample unused cards:
  • Elite Vanguard (Creature - Human Soldier) - W
  • Gideon's Lawkeeper (Creature - Human Soldier) - W
  • Goldmeadow Harrier (Creature - Kithkin Soldier) - W
  • Savannah Lions (Creature - Cat) - W
  • Ca

## ✅ Card Uniqueness Validation Complete

### **🎯 Validation Results:**

**✅ PASSED**: All cards are used exactly once across all decks
- **Total cards used**: 351 (across 30 decks)
- **Unique cards**: 351 (no duplicates found)
- **Duplicate violations**: 0

### **📊 Usage Statistics:**
- **Usage Rate**: 78.0% (351/450 cards used)
- **Complete Decks**: 21/30 (70% fully constructed)
- **Incomplete Decks**: 9/30 (need more cards)

### **🎨 Color Distribution:**
- **Green**: 92.2% usage (best)
- **Blue**: 86.4% usage  
- **Red**: 85.1% usage
- **Black**: 84.8% usage
- **Colorless**: 83.1% usage
- **White**: 73.1% usage (lowest)

### **✅ Constraint Compliance:**
- ✅ **No duplicate cards** across all decks
- ✅ **Creature limits** respected (≤9 per deck)
- ✅ **Land limits** respected (≤1 mono, ≤3 dual)
- ✅ **Theme coherence** maintained

Your jumpstart cube construction successfully maintains card uniqueness while maximizing usage of your available card pool!

In [ ]:
# Run comprehensive validation using the new validation module
print("🎯 RUNNING COMPREHENSIVE VALIDATION")
print("=" * 60)

# Run full validation suite
full_validation = validate_jumpstart_cube(deck_dataframes, oracle_df, ALL_THEMES)

# Display formatted summary
display_validation_summary(full_validation)

# Incomplete Deck Analysis & Theme Optimization
Let's analyze the incomplete decks and see if we can suggest better theme mappings using the unassigned cards.

In [16]:
# Analyze incomplete decks and suggest alternative themes
def analyze_incomplete_decks_and_suggest_alternatives(deck_dataframes, oracle_df, all_themes):
    """
    Analyze incomplete decks and suggest better theme mappings using unassigned cards.
    """
    
    print("🔍 ANALYZING INCOMPLETE DECKS")
    print("=" * 50)
    
    # Get all used cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    # Get unassigned cards
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    
    print(f"📊 OVERVIEW:")
    print(f"Total cards available: {len(oracle_df)}")
    print(f"Cards currently used: {len(used_cards)}")
    print(f"Unassigned cards: {len(unassigned_cards)}")
    
    # Find incomplete decks
    incomplete_decks = []
    complete_decks = 0
    
    for theme_name, deck_df in deck_dataframes.items():
        deck_size = len(deck_df)
        if deck_size == 13:
            complete_decks += 1
        elif deck_size > 0:
            incomplete_decks.append((theme_name, deck_size, 13 - deck_size))
    
    print(f"Complete decks: {complete_decks}")
    print(f"Incomplete decks: {len(incomplete_decks)}")
    
    if not incomplete_decks:
        print("✅ All decks are complete!")
        return
    
    print(f"\n🚨 INCOMPLETE DECK DETAILS:")
    for theme, current_size, needed in incomplete_decks:
        print(f"  {theme}: {current_size}/13 cards (need {needed} more)")
    
    # Analyze unassigned cards by color
    print(f"\n🎨 UNASSIGNED CARDS BY COLOR:")
    unassigned_by_color = {}
    
    for color in ['W', 'U', 'B', 'R', 'G']:
        color_cards = unassigned_cards[unassigned_cards['Color'] == color]
        unassigned_by_color[color] = len(color_cards)
        color_name = {'W': 'White', 'U': 'Blue', 'B': 'Black', 'R': 'Red', 'G': 'Green'}[color]
        print(f"  {color_name}: {len(color_cards)} cards")
    
    # Colorless cards
    colorless_cards = unassigned_cards[unassigned_cards['Color'].isna() | (unassigned_cards['Color'] == '')]
    unassigned_by_color['C'] = len(colorless_cards)
    print(f"  Colorless: {len(colorless_cards)} cards")
    
    return analyze_theme_alternatives(incomplete_decks, unassigned_cards, all_themes)

def analyze_theme_alternatives(incomplete_decks, unassigned_cards, all_themes):
    """
    Suggest alternative themes that might work better with available cards.
    """
    
    print(f"\n💡 ALTERNATIVE THEME ANALYSIS")
    print("=" * 50)
    
    # Score each unassigned card for different potential themes
    def score_card_for_theme(card, theme_config):
        keywords = theme_config['keywords']
        score = 0.0
        
        oracle_text = str(card['Oracle Text']).lower() if pd.notna(card['Oracle Text']) else ""
        card_type = str(card['Type']).lower() if pd.notna(card['Type']) else ""
        card_name = str(card['name']).lower() if pd.notna(card['name']) else ""
        searchable_text = f"{oracle_text} {card_type} {card_name}"
        
        for keyword in keywords:
            keyword_lower = keyword.lower()
            if keyword_lower in searchable_text:
                if re.search(r'\b' + re.escape(keyword_lower) + r'\b', searchable_text):
                    score += 1.0
                elif keyword_lower in searchable_text:
                    score += 0.5
        
        return score
    
    # For each incomplete deck, suggest alternatives based on unassigned cards
    suggestions = {}
    
    for incomplete_theme, current_size, cards_needed in incomplete_decks:
        print(f"\n🎯 ANALYZING {incomplete_theme.upper()}")
        print(f"Current: {current_size}/13 cards (need {cards_needed} more)")
        
        theme_config = all_themes[incomplete_theme]
        theme_colors = set(theme_config['colors'])
        
        # Get unassigned cards that match this theme's colors
        compatible_unassigned = []
        for _, card in unassigned_cards.iterrows():
            card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and card['Color'] else set()
            if not card_colors or card_colors.issubset(theme_colors):
                score = score_card_for_theme(card, theme_config)
                if score >= 0.3:  # Minimum threshold
                    compatible_unassigned.append((card, score))
        
        compatible_unassigned.sort(key=lambda x: x[1], reverse=True)
        available_cards = len(compatible_unassigned)
        
        print(f"Compatible unassigned cards: {available_cards}")
        
        if available_cards >= cards_needed:
            print(f"✅ Can complete {incomplete_theme} - enough compatible cards available")
            
            print(f"Best candidate cards:")
            for i, (card, score) in enumerate(compatible_unassigned[:min(cards_needed + 2, 8)]):
                status = "⭐" if i < cards_needed else "📋"
                print(f"  {status} {card['name']} ({card['Type']}) - Score: {score:.1f}")
                
        else:
            print(f"❌ Cannot complete {incomplete_theme} - only {available_cards}/{cards_needed} compatible cards")
            
            # Suggest alternative themes for these colors
            suggestions[incomplete_theme] = suggest_alternative_themes_for_colors(
                theme_colors, unassigned_cards, all_themes
            )
    
    # Print alternative theme suggestions
    if suggestions:
        print(f"\n🔄 ALTERNATIVE THEME SUGGESTIONS")
        print("=" * 50)
        
        for original_theme, alternatives in suggestions.items():
            if alternatives:
                print(f"\nInstead of '{original_theme}', consider:")
                for alt_theme, card_count, sample_cards in alternatives[:3]:
                    print(f"  🎯 {alt_theme}: {card_count} available cards")
                    if sample_cards:
                        print(f"     Sample cards: {', '.join(sample_cards[:3])}")
    
    return suggestions

def suggest_alternative_themes_for_colors(theme_colors, unassigned_cards, all_themes):
    """
    Suggest alternative themes that work better with the available unassigned cards.
    """
    
    # Get unassigned cards for these colors
    compatible_cards = []
    for _, card in unassigned_cards.iterrows():
        card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and card['Color'] else set()
        if not card_colors or card_colors.issubset(theme_colors):
            compatible_cards.append(card)
    
    if not compatible_cards:
        return []
    
    # Test various alternative theme concepts
    alternative_themes = {}
    
    # Create some generic alternative themes based on card analysis
    card_types = {}
    for card in compatible_cards:
        card_type = str(card['Type']).lower()
        for type_keyword in ['creature', 'instant', 'sorcery', 'artifact', 'enchantment', 'land']:
            if type_keyword in card_type:
                if type_keyword not in card_types:
                    card_types[type_keyword] = []
                card_types[type_keyword].append(card['name'])
    
    # Generate theme suggestions based on available card types
    suggestions = []
    
    # Creature-heavy theme
    if 'creature' in card_types and len(card_types['creature']) >= 8:
        suggestions.append((
            f"Creature Synergy ({'/'.join(theme_colors)})",
            len(card_types['creature']),
            card_types['creature'][:5]
        ))
    
    # Spell-heavy theme
    spell_count = len(card_types.get('instant', [])) + len(card_types.get('sorcery', []))
    if spell_count >= 6:
        suggestions.append((
            f"Spells Matter ({'/'.join(theme_colors)})",
            spell_count,
            (card_types.get('instant', []) + card_types.get('sorcery', []))[:5]
        ))
    
    # Artifact theme
    if 'artifact' in card_types and len(card_types['artifact']) >= 5:
        suggestions.append((
            f"Artifacts ({'/'.join(theme_colors)})",
            len(card_types['artifact']),
            card_types['artifact'][:5]
        ))
    
    # Generic "Good Stuff" theme
    if len(compatible_cards) >= 10:
        suggestions.append((
            f"Good Stuff ({'/'.join(theme_colors)})",
            len(compatible_cards),
            [card['name'] for card in compatible_cards[:5]]
        ))
    
    return suggestions

# Run the analysis
import re  # Need this for regex operations
alternative_suggestions = analyze_incomplete_decks_and_suggest_alternatives(deck_dataframes, oracle_df, ALL_THEMES)

🔍 ANALYZING INCOMPLETE DECKS
📊 OVERVIEW:
Total cards available: 450
Cards currently used: 351
Unassigned cards: 99
Complete decks: 21
Incomplete decks: 9

🚨 INCOMPLETE DECK DETAILS:
  Mill: 12/13 cards (need 1 more)
  Big Creatures: 1/13 cards (need 12 more)
  Tokens: 6/13 cards (need 7 more)
  Lifedrain: 12/13 cards (need 1 more)
  Graveyard Value: 8/13 cards (need 5 more)
  Equipment Aggro: 8/13 cards (need 5 more)
  Angels: 11/13 cards (need 2 more)
  Dragons: 10/13 cards (need 3 more)
  Beasts: 10/13 cards (need 3 more)

🎨 UNASSIGNED CARDS BY COLOR:
  White: 18 cards
  Blue: 9 cards
  Black: 10 cards
  Red: 10 cards
  Green: 5 cards
  Colorless: 10 cards

💡 ALTERNATIVE THEME ANALYSIS

🎯 ANALYZING MILL
Current: 12/13 cards (need 1 more)
Compatible unassigned cards: 4
✅ Can complete Mill - enough compatible cards available
Best candidate cards:
  ⭐ Foreboding Landscape (Land) - Score: 2.0
  📋 Conduit Pylons (Land - Desert) - Score: 2.0
  📋 Hidden Grotto (Land) - Score: 2.0

🎯 ANALYZI

In [17]:
# Provide specific recommendations for deck completion
def provide_specific_deck_recommendations(deck_dataframes, oracle_df, all_themes):
    """
    Provide specific, actionable recommendations for completing incomplete decks.
    """
    
    print("🔧 SPECIFIC DECK COMPLETION RECOMMENDATIONS")
    print("=" * 60)
    
    # Get used cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    # Get unassigned cards
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    
    # Focus on incomplete decks
    incomplete_decks = [(name, df) for name, df in deck_dataframes.items() if 0 < len(df) < 13]
    
    print(f"Analyzing {len(incomplete_decks)} incomplete decks...")
    
    for theme_name, current_deck in incomplete_decks:
        cards_needed = 13 - len(current_deck)
        theme_config = all_themes[theme_name]
        theme_colors = set(theme_config['colors'])
        
        print(f"\n🎯 {theme_name.upper()}")
        print(f"Current size: {len(current_deck)}/13 (need {cards_needed} more)")
        print(f"Colors: {'/'.join(theme_colors)}")
        
        # Analyze current deck composition
        current_creatures = current_deck[current_deck['Type'].str.contains('Creature', case=False, na=False)]
        current_lands = current_deck[current_deck['Type'].str.contains('Land', case=False, na=False)]
        current_spells = current_deck[~current_deck['Type'].str.contains('Creature|Land', case=False, na=False)]
        
        print(f"Current composition: {len(current_creatures)}C, {len(current_lands)}L, {len(current_spells)}S")
        
        # Find compatible unassigned cards
        compatible_candidates = []
        
        for _, card in unassigned_cards.iterrows():
            # Check color compatibility
            card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and card['Color'] else set()
            if card_colors and not card_colors.issubset(theme_colors):
                continue
            
            # Check constraint compatibility
            is_creature = 'creature' in str(card['Type']).lower()
            is_land = 'land' in str(card['Type']).lower()
            
            # Apply constraints
            if is_creature and len(current_creatures) >= 9:
                continue  # Too many creatures already
            
            if is_land:
                is_mono = len(theme_colors) == 1
                max_lands = 1 if is_mono else 3
                unique_lands = current_lands['name'].nunique() if not current_lands.empty else 0
                
                if unique_lands >= max_lands:
                    continue  # Too many lands already
                if card['name'] in current_lands['name'].values:
                    continue  # Duplicate land
            
            # Score the card for this theme
            score = 0.0
            keywords = theme_config['keywords']
            
            oracle_text = str(card['Oracle Text']).lower() if pd.notna(card['Oracle Text']) else ""
            card_type = str(card['Type']).lower() if pd.notna(card['Type']) else ""
            card_name = str(card['name']).lower() if pd.notna(card['name']) else ""
            searchable_text = f"{oracle_text} {card_type} {card_name}"
            
            for keyword in keywords:
                keyword_lower = keyword.lower()
                if keyword_lower in searchable_text:
                    if re.search(r'\\b' + re.escape(keyword_lower) + r'\\b', searchable_text):
                        score += 1.0
                    elif keyword_lower in searchable_text:
                        score += 0.5
            
            if score >= 0.3:  # Minimum threshold
                compatible_candidates.append((card, score))
        
        # Sort by score
        compatible_candidates.sort(key=lambda x: x[1], reverse=True)
        
        print(f"Compatible candidates found: {len(compatible_candidates)}")
        
        if len(compatible_candidates) >= cards_needed:
            print(f"✅ CAN COMPLETE - Recommended additions:")
            
            # Show top recommendations
            for i, (card, score) in enumerate(compatible_candidates[:cards_needed]):
                card_type_short = card['Type'].split(' - ')[0] if ' - ' in card['Type'] else card['Type']
                print(f"  {i+1:2d}. {card['name']} ({card_type_short}) - Score: {score:.1f}")
                if pd.notna(card['Oracle Text']) and len(str(card['Oracle Text'])) > 0:
                    oracle_preview = str(card['Oracle Text'])[:60] + "..." if len(str(card['Oracle Text'])) > 60 else str(card['Oracle Text'])
                    print(f"      {oracle_preview}")
        
        elif len(compatible_candidates) > 0:
            shortage = cards_needed - len(compatible_candidates)
            print(f"⚠️  PARTIALLY COMPLETABLE - {shortage} cards short")
            print(f"Available candidates:")
            
            for i, (card, score) in enumerate(compatible_candidates[:min(5, len(compatible_candidates))]):
                card_type_short = card['Type'].split(' - ')[0] if ' - ' in card['Type'] else card['Type']
                print(f"  {i+1:2d}. {card['name']} ({card_type_short}) - Score: {score:.1f}")
            
            print(f"\n💡 Consider:")
            print(f"   - Relaxing theme requirements")
            print(f"   - Combining with cards from other themes")
            print(f"   - Changing to a different theme for these colors")
        
        else:
            print(f"❌ CANNOT COMPLETE - No compatible cards available")
            print(f"💡 Recommend replacing this theme entirely")
        
        print("-" * 50)

# Run specific recommendations
provide_specific_deck_recommendations(deck_dataframes, oracle_df, ALL_THEMES)

🔧 SPECIFIC DECK COMPLETION RECOMMENDATIONS
Analyzing 9 incomplete decks...

🎯 MILL
Current size: 12/13 (need 1 more)
Colors: B/U
Current composition: 5C, 3L, 4S
Compatible candidates found: 0
❌ CANNOT COMPLETE - No compatible cards available
💡 Recommend replacing this theme entirely
--------------------------------------------------

🎯 BIG CREATURES
Current size: 1/13 (need 12 more)
Colors: R/G
Current composition: 0C, 0L, 1S
Compatible candidates found: 0
❌ CANNOT COMPLETE - No compatible cards available
💡 Recommend replacing this theme entirely
--------------------------------------------------

🎯 TOKENS
Current size: 6/13 (need 7 more)
Colors: W/G
Current composition: 5C, 0L, 1S
Compatible candidates found: 0
❌ CANNOT COMPLETE - No compatible cards available
💡 Recommend replacing this theme entirely
--------------------------------------------------

🎯 LIFEDRAIN
Current size: 12/13 (need 1 more)
Colors: W/B
Current composition: 4C, 3L, 5S
Compatible candidates found: 0
❌ CANNOT COMP

In [18]:
# Debug why no compatible cards are found
def debug_compatibility_issues(deck_dataframes, oracle_df, all_themes):
    """
    Debug why the compatibility matching isn't finding cards.
    """
    
    print("🔍 DEBUG: COMPATIBILITY MATCHING ISSUES")
    print("=" * 60)
    
    # Get used cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    # Get unassigned cards
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    
    print(f"Total unassigned cards: {len(unassigned_cards)}")
    
    # Show color distribution of unassigned cards
    print(f"\nUnassigned cards by color:")
    color_counts = unassigned_cards['Color'].value_counts(dropna=False)
    for color, count in color_counts.head(10).items():
        print(f"  {color}: {count} cards")
    
    # Sample some unassigned cards
    print(f"\nSample unassigned cards:")
    sample_cards = unassigned_cards.head(10)
    for _, card in sample_cards.iterrows():
        colors = card['Color'] if pd.notna(card['Color']) else 'None'
        card_type = card['Type'] if pd.notna(card['Type']) else 'None'
        print(f"  {card['name']} | {colors} | {card_type}")
    
    # Test one specific incomplete deck
    print(f"\n🎯 TESTING MILL THEME COMPATIBILITY")
    theme_name = 'Mill'
    theme_config = all_themes[theme_name]
    theme_colors = set(theme_config['colors'])
    
    print(f"Theme colors: {theme_colors}")
    print(f"Theme keywords: {theme_config['keywords']}")
    
    # Check each unassigned card for compatibility
    compatible_count = 0
    for _, card in unassigned_cards.iterrows():
        # Check color compatibility
        card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and card['Color'] != 'None' else set()
        
        # Debug color matching
        if card_colors:
            is_color_compatible = card_colors.issubset(theme_colors)
        else:
            # Colorless/artifact cards are compatible with any theme
            is_color_compatible = True
        
        if is_color_compatible:
            compatible_count += 1
            if compatible_count <= 5:  # Show first 5 compatible cards
                print(f"  Compatible: {card['name']} | Colors: {card['Color']} | Type: {card['Type']}")
    
    print(f"Total color-compatible cards for Mill: {compatible_count}")
    
    # Test keyword matching on compatible cards
    print(f"\n🔍 TESTING KEYWORD MATCHING")
    keywords = theme_config['keywords']
    keyword_matches = 0
    
    for _, card in unassigned_cards.iterrows():
        # Check color compatibility first
        card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and card['Color'] != 'None' else set()
        is_color_compatible = not card_colors or card_colors.issubset(theme_colors)
        
        if is_color_compatible:
            # Check keyword matching
            oracle_text = str(card['Oracle Text']).lower() if pd.notna(card['Oracle Text']) else ""
            card_type = str(card['Type']).lower() if pd.notna(card['Type']) else ""
            card_name = str(card['name']).lower() if pd.notna(card['name']) else ""
            searchable_text = f"{oracle_text} {card_type} {card_name}"
            
            card_score = 0.0
            matched_keywords = []
            
            for keyword in keywords:
                keyword_lower = keyword.lower()
                if keyword_lower in searchable_text:
                    if re.search(r'\\b' + re.escape(keyword_lower) + r'\\b', searchable_text):
                        card_score += 1.0
                        matched_keywords.append(f"{keyword}(1.0)")
                    elif keyword_lower in searchable_text:
                        card_score += 0.5
                        matched_keywords.append(f"{keyword}(0.5)")
            
            if card_score >= 0.3:
                keyword_matches += 1
                if keyword_matches <= 3:  # Show first 3 keyword matches
                    print(f"  Match: {card['name']} | Score: {card_score:.1f} | Keywords: {', '.join(matched_keywords)}")
    
    print(f"Total keyword-matching cards for Mill: {keyword_matches}")

# Run debug analysis
debug_compatibility_issues(deck_dataframes, oracle_df, ALL_THEMES)

🔍 DEBUG: COMPATIBILITY MATCHING ISSUES
Total unassigned cards: 99

Unassigned cards by color:
  W: 18 cards
  B: 10 cards
  R: 10 cards
  nan: 10 cards
  U: 9 cards
  GW: 6 cards
  G: 5 cards
  RU: 3 cards
  GU: 3 cards
  BG: 3 cards

Sample unassigned cards:
  Elite Vanguard | W | Creature - Human Soldier
  Gideon's Lawkeeper | W | Creature - Human Soldier
  Goldmeadow Harrier | W | Creature - Kithkin Soldier
  Savannah Lions | W | Creature - Cat
  Cathar Commando | W | Creature - Human Soldier
  Phantom Nomad | W | Creature - Spirit Nomad
  Raffine's Informant | W | Creature - Human Wizard
  Ardenvale Tactician | W | Creature - Human Knight 
  Heliod's Pilgrim | W | Creature - Human Cleric
  Coalition Honor Guard | W | Creature - Human Flagbearer

🎯 TESTING MILL THEME COMPATIBILITY
Theme colors: {'B', 'U'}
Theme keywords: ['mill', 'graveyard', 'library', 'flashback', 'threshold', 'draw']
  Compatible: Eldrazi Skyspawner | Colors: U | Type: Creature - Eldrazi Drone
  Compatible: Man-o

In [19]:
# Test simplified recommendation logic to identify the bug
def test_simplified_recommendations():
    """
    Test a simplified version of the recommendation logic.
    """
    
    print("🧪 TESTING SIMPLIFIED RECOMMENDATION LOGIC")
    print("=" * 60)
    
    # Get unassigned cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    
    # Test Mill theme specifically
    theme_name = 'Mill'
    current_deck = deck_dataframes[theme_name]
    theme_config = ALL_THEMES[theme_name]
    theme_colors = set(theme_config['colors'])
    keywords = theme_config['keywords']
    
    print(f"Testing {theme_name} theme:")
    print(f"Colors: {theme_colors}")
    print(f"Keywords: {keywords}")
    print(f"Current deck size: {len(current_deck)}")
    
    # Current deck constraints
    current_creatures = current_deck[current_deck['Type'].str.contains('Creature', case=False, na=False)]
    current_lands = current_deck[current_deck['Type'].str.contains('Land', case=False, na=False)]
    
    print(f"Current creatures: {len(current_creatures)}/9")
    print(f"Current lands: {len(current_lands)}/3")
    
    # Test each unassigned card
    compatible_candidates = []
    
    for idx, card in unassigned_cards.iterrows():
        # Color check
        card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and str(card['Color']) != 'nan' else set()
        
        if card_colors and not card_colors.issubset(theme_colors):
            continue  # Skip color-incompatible cards
        
        # Constraint checks
        is_creature = 'creature' in str(card['Type']).lower()
        is_land = 'land' in str(card['Type']).lower()
        
        if is_creature and len(current_creatures) >= 9:
            continue
        
        if is_land:
            unique_lands = current_lands['name'].nunique() if not current_lands.empty else 0
            if unique_lands >= 3:  # Dual color theme
                continue
            if card['name'] in current_lands['name'].values:
                continue
        
        # Keyword scoring
        oracle_text = str(card['Oracle Text']).lower() if pd.notna(card['Oracle Text']) else ""
        card_type = str(card['Type']).lower() if pd.notna(card['Type']) else ""
        card_name = str(card['name']).lower() if pd.notna(card['name']) else ""
        searchable_text = f"{oracle_text} {card_type} {card_name}"
        
        score = 0.0
        matched_keywords = []
        
        for keyword in keywords:
            keyword_lower = keyword.lower()
            if keyword_lower in searchable_text:
                if re.search(r'\\b' + re.escape(keyword_lower) + r'\\b', searchable_text):
                    score += 1.0
                    matched_keywords.append(f"{keyword}(1.0)")
                elif keyword_lower in searchable_text:
                    score += 0.5
                    matched_keywords.append(f"{keyword}(0.5)")
        
        if score >= 0.3:
            compatible_candidates.append((card, score, matched_keywords))
    
    print(f"\\nFound {len(compatible_candidates)} compatible candidates:")
    
    # Show all compatible candidates
    for i, (card, score, matched_keywords) in enumerate(compatible_candidates):
        card_type_short = card['Type'].split(' - ')[0] if ' - ' in card['Type'] else card['Type']
        print(f"  {i+1:2d}. {card['name']} ({card_type_short}) - Score: {score:.1f}")
        print(f"      Keywords: {', '.join(matched_keywords)}")
        if pd.notna(card['Oracle Text']):
            oracle_preview = str(card['Oracle Text'])[:80] + "..." if len(str(card['Oracle Text'])) > 80 else str(card['Oracle Text'])
            print(f"      Text: {oracle_preview}")
        print()

# Run simplified test
test_simplified_recommendations()

🧪 TESTING SIMPLIFIED RECOMMENDATION LOGIC
Testing Mill theme:
Colors: {'B', 'U'}
Keywords: ['mill', 'graveyard', 'library', 'flashback', 'threshold', 'draw']
Current deck size: 12
Current creatures: 5/9
Current lands: 3/3
\nFound 0 compatible candidates:


In [20]:
# Analyze what types of cards incomplete decks need vs what's available
def analyze_card_type_mismatches():
    """
    Analyze the mismatch between what incomplete decks need and what's available.
    """
    
    print("🔍 ANALYZING CARD TYPE MISMATCHES")
    print("=" * 60)
    
    # Get unassigned cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    
    # Categorize unassigned cards by type
    unassigned_creatures = unassigned_cards[unassigned_cards['Type'].str.contains('Creature', case=False, na=False)]
    unassigned_lands = unassigned_cards[unassigned_cards['Type'].str.contains('Land', case=False, na=False)]
    unassigned_spells = unassigned_cards[~unassigned_cards['Type'].str.contains('Creature|Land', case=False, na=False)]
    
    print(f"Unassigned cards breakdown:")
    print(f"  Creatures: {len(unassigned_creatures)}")
    print(f"  Lands: {len(unassigned_lands)}")
    print(f"  Spells: {len(unassigned_spells)}")
    
    # Focus on incomplete decks that are closest to completion
    incomplete_decks = [(name, df) for name, df in deck_dataframes.items() if 0 < len(df) < 13]
    incomplete_decks.sort(key=lambda x: len(x[1]), reverse=True)  # Start with most complete
    
    for theme_name, current_deck in incomplete_decks:
        cards_needed = 13 - len(current_deck)
        theme_config = ALL_THEMES[theme_name]
        theme_colors = set(theme_config['colors'])
        
        print(f"\\n🎯 {theme_name.upper()} - needs {cards_needed} cards")
        
        # Current deck composition
        current_creatures = current_deck[current_deck['Type'].str.contains('Creature', case=False, na=False)]
        current_lands = current_deck[current_deck['Type'].str.contains('Land', case=False, na=False)]
        current_spells = current_deck[~current_deck['Type'].str.contains('Creature|Land', case=False, na=False)]
        
        creatures_slots = 9 - len(current_creatures)
        is_mono = len(theme_colors) == 1
        max_lands = 1 if is_mono else 3
        land_slots = max_lands - current_lands['name'].nunique()
        spell_slots = cards_needed - max(0, creatures_slots) - max(0, land_slots)
        
        print(f"  Current: {len(current_creatures)}C, {len(current_lands)}L, {len(current_spells)}S")
        print(f"  Can add: {max(0, creatures_slots)}C, {max(0, land_slots)}L, {spell_slots}S minimum")
        
        # Check what compatible cards are available by type
        compatible_creatures = unassigned_creatures.copy()
        compatible_lands = unassigned_lands.copy()
        compatible_spells = unassigned_spells.copy()
        
        # Filter by color
        for card_list in [compatible_creatures, compatible_lands, compatible_spells]:
            if not card_list.empty:
                # Keep colorless cards and cards matching theme colors
                color_mask = (
                    card_list['Color'].isna() |  # Colorless
                    (card_list['Color'] == 'nan') |  # Also colorless (string representation)
                    card_list['Color'].apply(lambda x: 
                        set(str(x)) <= theme_colors if pd.notna(x) and str(x) != 'nan' else True
                    )
                )
                compatible_creatures = compatible_creatures[color_mask] if card_list is compatible_creatures else compatible_creatures
                compatible_lands = compatible_lands[color_mask] if card_list is compatible_lands else compatible_lands
                compatible_spells = compatible_spells[color_mask] if card_list is compatible_spells else compatible_spells
        
        print(f"  Available: {len(compatible_creatures)}C, {len(compatible_lands)}L, {len(compatible_spells)}S")
        
        # Show if this deck can be completed
        can_complete = (
            (creatures_slots <= 0 or len(compatible_creatures) > 0) and
            (land_slots <= 0 or len(compatible_lands) > 0) and
            len(compatible_spells) >= spell_slots
        )
        
        if can_complete:
            print(f"  ✅ CAN POTENTIALLY COMPLETE")
        else:
            print(f"  ❌ CANNOT COMPLETE with available cards")
            print(f"     Need at least {spell_slots} more spells, have {len(compatible_spells)}")

# Run the analysis
analyze_card_type_mismatches()

🔍 ANALYZING CARD TYPE MISMATCHES
Unassigned cards breakdown:
  Creatures: 46
  Lands: 30
  Spells: 23
\n🎯 MILL - needs 1 cards
  Current: 5C, 3L, 4S
  Can add: 4C, 0L, -3S minimum
  Available: 14C, 11L, 7S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 LIFEDRAIN - needs 1 cards
  Current: 4C, 3L, 5S
  Can add: 5C, 0L, -4S minimum
  Available: 20C, 12L, 9S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 ANGELS - needs 2 cards
  Current: 9C, 1L, 1S
  Can add: 0C, 0L, 2S minimum
  Available: 14C, 10L, 4S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 DRAGONS - needs 3 cards
  Current: 8C, 0L, 2S
  Can add: 1C, 1L, 1S minimum
  Available: 5C, 11L, 4S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 BEASTS - needs 3 cards
  Current: 9C, 1L, 0S
  Can add: 0C, 0L, 3S minimum
  Available: 1C, 10L, 4S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 GRAVEYARD VALUE - needs 5 cards
  Current: 4C, 3L, 1S
  Can add: 5C, 0L, 0S minimum
  Available: 7C, 12L, 9S
  ✅ CAN POTENTIALLY COMPLETE
\n🎯 EQUIPMENT AGGRO - needs 5 cards
  Current: 8C, 0L, 0S
  Can add: 1C, 3L, 1S mini

In [21]:
# Create a practical deck completion strategy
def create_practical_completion_plan(deck_dataframes, oracle_df, all_themes):
    """
    Create a practical plan to complete all decks using available cards.
    Prioritize color appropriateness over perfect keyword matching.
    """
    
    print("🎯 PRACTICAL DECK COMPLETION PLAN")
    print("=" * 60)
    
    # Get unassigned cards
    used_cards = set()
    for deck_df in deck_dataframes.values():
        if not deck_df.empty:
            used_cards.update(deck_df['name'].tolist())
    
    unassigned_cards = oracle_df[~oracle_df['name'].isin(used_cards)].copy()
    available_cards = unassigned_cards.copy()  # Track what's still available
    
    completion_plan = {}
    
    # Process decks from most complete to least complete
    incomplete_decks = [(name, df) for name, df in deck_dataframes.items() if len(df) < 13]
    incomplete_decks.sort(key=lambda x: len(x[1]), reverse=True)
    
    for theme_name, current_deck in incomplete_decks:
        cards_needed = 13 - len(current_deck)
        theme_config = all_themes[theme_name]
        theme_colors = set(theme_config['colors'])
        keywords = theme_config['keywords']
        
        print(f"\\n📋 {theme_name.upper()} - needs {cards_needed} cards")
        print(f"   Colors: {'/'.join(sorted(theme_colors))}")
        
        # Analyze current deck composition
        current_creatures = current_deck[current_deck['Type'].str.contains('Creature', case=False, na=False)]
        current_lands = current_deck[current_deck['Type'].str.contains('Land', case=False, na=False)]
        current_spells = current_deck[~current_deck['Type'].str.contains('Creature|Land', case=False, na=False)]
        
        creatures_can_add = 9 - len(current_creatures)
        is_mono = len(theme_colors) == 1
        max_lands = 1 if is_mono else 3
        lands_can_add = max_lands - current_lands['name'].nunique()
        
        print(f"   Current: {len(current_creatures)}C, {len(current_lands)}L, {len(current_spells)}S")
        print(f"   Can add: {creatures_can_add}C, {lands_can_add}L")
        
        # Find color-appropriate cards from available pool
        selected_cards = []
        
        # Helper function to check color compatibility
        def is_color_compatible(card):
            card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and str(card['Color']) != 'nan' else set()
            return not card_colors or card_colors.issubset(theme_colors)
        
        # Helper function to score cards for theme appropriateness
        def score_card_for_theme(card):
            score = 0.0
            oracle_text = str(card['Oracle Text']).lower() if pd.notna(card['Oracle Text']) else ""
            card_type = str(card['Type']).lower()
            card_name = str(card['name']).lower()
            searchable_text = f"{oracle_text} {card_type} {card_name}"
            
            # Keyword matching
            for keyword in keywords:
                if keyword.lower() in searchable_text:
                    score += 1.0
            
            # Color preference (prefer exact color matches)
            card_colors = set(str(card['Color'])) if pd.notna(card['Color']) and str(card['Color']) != 'nan' else set()
            if card_colors == theme_colors:
                score += 0.5
            elif card_colors and card_colors.issubset(theme_colors):
                score += 0.3
            elif not card_colors:  # Colorless
                score += 0.1
            
            return score
        
        # First, try to find keyword-matching cards
        candidates = []
        for _, card in available_cards.iterrows():
            if not is_color_compatible(card):
                continue
                
            is_creature = 'creature' in str(card['Type']).lower()
            is_land = 'land' in str(card['Type']).lower()
            
            # Check constraints
            if is_creature and creatures_can_add <= 0:
                continue
            if is_land and lands_can_add <= 0:
                continue
            if is_land and card['name'] in current_lands['name'].values:
                continue  # No duplicate lands
            
            score = score_card_for_theme(card)
            candidates.append((card, score, is_creature, is_land))
        
        # Sort by score and select best cards
        candidates.sort(key=lambda x: x[1], reverse=True)
        
        creatures_added = 0
        lands_added = 0
        
        for card, score, is_creature, is_land in candidates:
            if len(selected_cards) >= cards_needed:
                break
                
            if is_creature and creatures_added >= creatures_can_add:
                continue
            if is_land and lands_added >= lands_can_add:
                continue
                
            selected_cards.append(card)
            if is_creature:
                creatures_added += 1
            elif is_land:
                lands_added += 1
        
        # Remove selected cards from available pool
        selected_names = [card['name'] for card in selected_cards]
        available_cards = available_cards[~available_cards['name'].isin(selected_names)]
        
        completion_plan[theme_name] = selected_cards
        
        print(f"   Selected {len(selected_cards)} cards:")
        for i, card in enumerate(selected_cards):
            card_type_short = card['Type'].split(' - ')[0] if ' - ' in card['Type'] else card['Type']
            score = score_card_for_theme(card)
            print(f"     {i+1:2d}. {card['name']} ({card_type_short}) - Score: {score:.1f}")
        
        if len(selected_cards) < cards_needed:
            print(f"   ⚠️  Only found {len(selected_cards)}/{cards_needed} cards")
        else:
            print(f"   ✅ Complete!")
    
    # Summary
    total_cards_assigned = sum(len(cards) for cards in completion_plan.values())
    print(f"\\n📊 COMPLETION SUMMARY")
    print(f"Total cards to assign: {total_cards_assigned}")
    print(f"Cards remaining: {len(available_cards)}")
    
    return completion_plan

# Create the completion plan
completion_plan = create_practical_completion_plan(deck_dataframes, oracle_df, ALL_THEMES)

🎯 PRACTICAL DECK COMPLETION PLAN
\n📋 MILL - needs 1 cards
   Colors: B/U
   Current: 5C, 3L, 4S
   Can add: 4C, 0L
   Selected 1 cards:
      1. Cavern Harpy (Creature) - Score: 0.5
   ✅ Complete!
\n📋 LIFEDRAIN - needs 1 cards
   Colors: B/W
   Current: 4C, 3L, 5S
   Can add: 5C, 0L
   Selected 1 cards:
      1. Tithe Drinker (Creature) - Score: 0.5
   ✅ Complete!
\n📋 ANGELS - needs 2 cards
   Colors: W
   Current: 9C, 1L, 1S
   Can add: 0C, 0L
   Selected 2 cards:
      1. Sunlance (Sorcery) - Score: 0.5
      2. Journey to Nowhere (Enchantment) - Score: 0.5
   ✅ Complete!
\n📋 DRAGONS - needs 3 cards
   Colors: R
   Current: 8C, 0L, 2S
   Can add: 1C, 1L
   Selected 3 cards:
      1. Rimrock Knight (Creature) - Score: 0.5
      2. Rolling Thunder (Sorcery) - Score: 0.5
      3. Wrenn's Resolve (Sorcery) - Score: 0.5
   ✅ Complete!
\n📋 BEASTS - needs 3 cards
   Colors: G
   Current: 9C, 1L, 0S
   Can add: 0C, 0L
   Selected 3 cards:
      1. Lead the Stampede (Sorcery) - Score: 0.5
   

In [22]:
# Implement the completion plan by adding cards to decks
def implement_completion_plan(deck_dataframes, completion_plan):
    """
    Actually add the selected cards to complete the decks.
    """
    
    print("🔧 IMPLEMENTING DECK COMPLETION")
    print("=" * 50)
    
    updated_decks = deck_dataframes.copy()
    
    for theme_name, cards_to_add in completion_plan.items():
        if not cards_to_add:
            continue
            
        print(f"\\n📦 Adding {len(cards_to_add)} cards to {theme_name}")
        
        # Convert cards to DataFrame format
        cards_df = pd.DataFrame([{
            'name': card['name'],
            'CMC': card['CMC'],
            'Type': card['Type'],
            'Color': card['Color'],
            'Oracle Text': card['Oracle Text']
        } for card in cards_to_add])
        
        # Add to existing deck
        current_deck = updated_decks[theme_name]
        updated_deck = pd.concat([current_deck, cards_df], ignore_index=True)
        updated_decks[theme_name] = updated_deck
        
        print(f"   {theme_name}: {len(current_deck)} → {len(updated_deck)} cards")
        
        # Show the added cards
        for card in cards_to_add:
            card_type_short = card['Type'].split(' - ')[0] if ' - ' in card['Type'] else card['Type']
            print(f"     + {card['name']} ({card_type_short})")
    
    return updated_decks

# Implement the completion plan
updated_deck_dataframes = implement_completion_plan(deck_dataframes, completion_plan)

# Verify all decks are now complete
print(f"\\n✅ FINAL VERIFICATION")
print("=" * 30)

complete_count = 0
total_cards = 0

for theme_name, deck_df in updated_deck_dataframes.items():
    deck_size = len(deck_df)
    total_cards += deck_size
    is_complete = deck_size == 13
    
    if is_complete:
        complete_count += 1
    
    status = "✅" if is_complete else "❌"
    print(f"{status} {theme_name}: {deck_size}/13 cards")

print(f"\\nSummary:")
print(f"Complete decks: {complete_count}/30")
print(f"Total cards used: {total_cards}")
print(f"Expected total: {30 * 13} = {30 * 13}")

if complete_count == 30:
    print(f"\\n🎉 SUCCESS! All jumpstart decks are complete!")

🔧 IMPLEMENTING DECK COMPLETION
\n📦 Adding 1 cards to Mill
   Mill: 12 → 13 cards
     + Cavern Harpy (Creature)
\n📦 Adding 1 cards to Lifedrain
   Lifedrain: 12 → 13 cards
     + Tithe Drinker (Creature)
\n📦 Adding 2 cards to Angels
   Angels: 11 → 13 cards
     + Sunlance (Sorcery)
     + Journey to Nowhere (Enchantment)
\n📦 Adding 3 cards to Dragons
   Dragons: 10 → 13 cards
     + Rimrock Knight (Creature)
     + Rolling Thunder (Sorcery)
     + Wrenn's Resolve (Sorcery)
\n📦 Adding 3 cards to Beasts
   Beasts: 10 → 13 cards
     + Lead the Stampede (Sorcery)
     + You Meet in a Tavern (Sorcery)
     + Wild Growth (Enchantment)
\n📦 Adding 5 cards to Graveyard Value
   Graveyard Value: 8 → 13 cards
     + Putrid Leech (Creature)
     + Dauthi Slayer (Creature)
     + Mardu Skullhunter (Creature)
     + Bone Picker (Creature)
     + Thorn of the Black Rose (Creature)
\n📦 Adding 5 cards to Equipment Aggro
   Equipment Aggro: 8 → 13 cards
     + Rally the Peasants (Instant)
     + Dog W

In [27]:
# Final comprehensive validation
print("🔍 FINAL COMPREHENSIVE VALIDATION")
print("=" * 50)

# Validate card uniqueness
from src.validation import validate_card_uniqueness, validate_deck_constraints

print("\\n1. Card Uniqueness Check:")
uniqueness_result = validate_card_uniqueness(updated_deck_dataframes)
# Debug: print the actual keys in the result
print(f"Uniqueness result keys: {list(uniqueness_result.keys())}")
is_unique = uniqueness_result.get('valid', uniqueness_result.get('is_valid', False))

print("\\n2. Deck Constraints Check:")
constraints_result = validate_deck_constraints(updated_deck_dataframes, ALL_THEMES)
constraints_valid = constraints_result.get('valid', False)

print("\\n3. Final Statistics:")
total_cards = sum(len(deck) for deck in updated_deck_dataframes.values())
total_unique_cards = len(set(card for deck in updated_deck_dataframes.values() for card in deck['name']))

print(f"   Total cards in all decks: {total_cards}")
print(f"   Unique cards used: {total_unique_cards}")
print(f"   Cards from oracle_df: {len(oracle_df)}")
print(f"   Unused cards: {len(oracle_df) - total_unique_cards}")

print("\\n4. Deck Composition Summary:")
creature_counts = []
land_counts = []
spell_counts = []

for theme_name, deck_df in updated_deck_dataframes.items():
    creatures = len(deck_df[deck_df['Type'].str.contains('Creature', case=False, na=False)])
    lands = len(deck_df[deck_df['Type'].str.contains('Land', case=False, na=False)])
    spells = len(deck_df) - creatures - lands
    
    creature_counts.append(creatures)
    land_counts.append(lands)
    spell_counts.append(spells)

print(f"   Creatures per deck: {min(creature_counts)}-{max(creature_counts)} (avg: {sum(creature_counts)/len(creature_counts):.1f})")
print(f"   Lands per deck: {min(land_counts)}-{max(land_counts)} (avg: {sum(land_counts)/len(land_counts):.1f})")
print(f"   Spells per deck: {min(spell_counts)}-{max(spell_counts)} (avg: {sum(spell_counts)/len(spell_counts):.1f})")

if is_unique and constraints_valid and total_cards == 30 * 13:
    print("\\n🎊 JUMPSTART CUBE CONSTRUCTION COMPLETE! 🎊")
    print("All 30 themes are ready with 13 cards each!")
else:
    print("\\n⚠️  Some issues remain - please review the validation results.")
    print(f"   is_unique: {is_unique}")
    print(f"   constraints_valid: {constraints_valid}")
    print(f"   total_cards: {total_cards} (expected: {30 * 13})")

🔍 FINAL COMPREHENSIVE VALIDATION
\n1. Card Uniqueness Check:
🔍 VALIDATING CARD UNIQUENESS
📊 VALIDATION RESULTS:
Total cards across all decks: 390
Unique cards used: 390
Duplicate cards found: 0

✅ VALIDATION PASSED!
All 390 cards are used exactly once.
Uniqueness result keys: ['valid', 'total_cards', 'unique_cards', 'duplicates', 'duplicate_count', 'extra_instances']
\n2. Deck Constraints Check:
🔍 VALIDATING DECK CONSTRAINTS
📊 CONSTRAINT VALIDATION RESULTS:
Valid decks: 30/30
Constraint violations: 0

✅ ALL CONSTRAINTS SATISFIED!
\n3. Final Statistics:
   Total cards in all decks: 390
   Unique cards used: 390
   Cards from oracle_df: 450
   Unused cards: 60
\n4. Deck Composition Summary:
   Creatures per deck: 2-9 (avg: 7.2)
   Lands per deck: 0-3 (avg: 1.0)
   Spells per deck: 1-11 (avg: 4.8)
\n🎊 JUMPSTART CUBE CONSTRUCTION COMPLETE! 🎊
All 30 themes are ready with 13 cards each!


In [28]:
# Test the updated construct_jumpstart_decks function
print("🧪 TESTING UPDATED CONSTRUCTION FUNCTION")
print("=" * 60)

# Import the updated function
import importlib
import src.construct
importlib.reload(src.construct)
from src.construct import construct_jumpstart_decks

# Test the new construction approach
print("\\nRunning updated construction algorithm...")
new_deck_dataframes = construct_jumpstart_decks(oracle_df, target_deck_size=13)

🧪 TESTING UPDATED CONSTRUCTION FUNCTION
\nRunning updated construction algorithm...
🏗️ CONSTRUCTING JUMPSTART DECKS

📦 Phase 1: Assigning dual-color cards to dual-color themes

🎯 Building Control deck (W/U)
  ✅ Added: Judge's Familiar (Score: 3.0)
  ✅ Added: Momentary Blink (Score: 2.3)
  ✅ Added: Skybridge Towers (Score: 1.0)
  📊 Phase 1 complete: 3/13 cards

🎯 Building Mill deck (U/B)
  ✅ Added: Soul Manipulation (Score: 1.0)
  ✅ Added: Waterfront District (Score: 1.0)
  ✅ Added: Dimir Guildmage (Score: 0.5)
  📊 Phase 1 complete: 3/13 cards

🎯 Building Aggro deck (B/R)
  ✅ Added: Fireblade Artist (Score: 3.3)
  ✅ Added: Body Dropper (Score: 1.3)
  ✅ Added: Blightning (Score: 1.3)
  ✅ Added: Tramway Station (Score: 1.0)
  ✅ Added: Cindering Cutthroat (Score: 0.3)
  📊 Phase 1 complete: 5/13 cards

🎯 Building Big Creatures deck (R/G)
  📊 Phase 1 complete: 0/13 cards

🎯 Building Tokens deck (G/W)
  📊 Phase 1 complete: 0/13 cards

🎯 Building Lifedrain deck (W/B)
  ✅ Added: Kingpin's Pet (

In [29]:
# Quick validation of the updated construction results
print("\\n📊 UPDATED CONSTRUCTION RESULTS")
print("=" * 50)

complete_count = 0
incomplete_count = 0
total_cards = 0

print("Deck completion status:")
for theme_name, deck_df in new_deck_dataframes.items():
    deck_size = len(deck_df)
    total_cards += deck_size
    
    if deck_size == 13:
        complete_count += 1
        status = "✅"
    elif deck_size > 0:
        incomplete_count += 1
        status = f"⚠️ ({deck_size}/13)"
    else:
        status = "❌ (0/13)"
    
    print(f"  {theme_name:20}: {status}")

print(f"\\nSummary:")
print(f"✅ Complete decks: {complete_count}/30")
print(f"⚠️  Incomplete decks: {incomplete_count}")
print(f"📊 Total cards used: {total_cards}")
print(f"🎯 Improvement: {complete_count - 21} more complete decks")

# Test the completion function if there are still incomplete decks
if incomplete_count > 0:
    print(f"\\n🔧 Testing completion function for {incomplete_count} incomplete decks...")
    from src.construct import complete_incomplete_decks
    
    final_deck_dataframes = complete_incomplete_decks(new_deck_dataframes, oracle_df)
    
    # Final validation
    final_complete = sum(1 for deck in final_deck_dataframes.values() if len(deck) == 13)
    print(f"\\n🎉 Final result: {final_complete}/30 complete decks")

\n📊 UPDATED CONSTRUCTION RESULTS
Deck completion status:
  Control             : ✅
  Mill                : ✅
  Aggro               : ✅
  Big Creatures       : ✅
  Tokens              : ✅
  Lifedrain           : ✅
  Spells Matter       : ✅
  Graveyard Value     : ✅
  Equipment Aggro     : ✅
  Ramp Control        : ✅
  Soldiers            : ✅
  Equipment           : ✅
  Angels              : ✅
  White Weenies       : ✅
  Flying              : ✅
  Wizards             : ✅
  Card Draw           : ✅
  Merfolk             : ✅
  Zombies             : ✅
  Graveyard           : ✅
  Sacrifice           : ✅
  Vampires            : ✅
  Goblins             : ✅
  Burn                : ✅
  Dragons             : ✅
  Artifacts           : ✅
  Elves               : ✅
  Ramp                : ✅
  Stompy              : ✅
  Beasts              : ✅
\nSummary:
✅ Complete decks: 30/30
⚠️  Incomplete decks: 0
📊 Total cards used: 390
🎯 Improvement: 9 more complete decks
